### DISCUSSION TITLE
ZSH integrations not working

### DISCUSSION THREAD

**Discussion by ldelossa** [kovidgoyal/kitty#5296] [Q&A] [answered]
Hey there.

For some reason my zsh integrations have stopped working. I can no longer jump to previous shell prompts. 
Anyway I can troubleshoot this?

```
zsh --version
zsh 5.8.1 (x86_64-redhat-linux-gnu)
```

```
kitty --version
kitty 0.25.2 created by Kovid Goyal
```

kitty conf
```
shell zsh

mouse_hide_wait 0
allow_remote_control yes

window_padding_width 2

font_family Cousine Medium
font_size   12.5

# these are used by fzf in vim
map ctrl+shift+v no_op
map ctrl+shift+s no_op
map ctrl+shift+t no_op
map ctrl+shit+j no_op 
map ctrl+shit+k no_op
map ctrl+shit+n no_op
map ctrl+shit+w no_op
map ctrl+shit+t no_op
map ctrl+shit+p no_op
map ctrl+shit+n no_op
map ctrl+shit+h no_op
map ctrl+shit+g no_op
map ctrl+shit+v no_op

map alt+shit+p no_op

map ctrl+c copy_or_interrupt
map ctrl+v paste_from_clipboard
map ctrl+alt+j scroll_line_down
map ctrl+alt+k scroll_line_up
map ctrl+alt+w close_window
map ctrl+alt+t new_tab
map ctrl+alt+p scroll_to_prompt -1
map ctrl+alt+n scroll_to_prompt 1
map ctrl+alt+h show_scrollback
map ctrl+alt+g show_last_command_output
map ctrl+alt+v show_last_visited_command_output
map ctrl+alt+u scroll_page_up
map ctrl+alt+d scroll_page_down

enable_audio_bell no

copy_on_select clipboard
skip_trailing_lines smart

mouse_hide_wait 0
```

Essentially I'm expecting "ctrl+alt+p" and "ctrl+alt+n" to move my kitty terminal's cursor up and down between shell prompts. Unfortunately this does not appear to work.

**Comment by kovidgoyal** [comment]
Start with empty kitty.conf and empty zshrc/zsenv/etc and put them back
in sections to find the culprit.

**Reply by ldelossa** [reply]
I now have a zshrc that only consists of 

```
# BEGIN_KITTY_SHELL_INTEGRATION
if test -e "/usr/lib64/kitty/shell-integration/zsh/kitty.zsh"; then source "/usr/lib64/kitty/shell-integration/zsh/kitty.zsh"; fi
# END_KITTY_SHELL_INTEGRATION
```

and a completely empty kitty.conf 

the default `ctrl-shift-z/x` still do not work.

**Comment by kovidgoyal** [comment]
Remove the KITTY_SHELL_INTEGRATION section it is not required.

**Comment by ldelossa** [comment]
still no success.

**Reply by kovidgoyal** [reply]
I dont see how that's possible. You are not doing something you think you are doing.

**Reply by ldelossa** [reply]
Well, i completely zero'd out both configs. There's no other settings to remove. Is there any further way to debug this? Can I send the command directly to kitty? Is there a way to see if the necessary functions exist in ZSH?

**Reply by kovidgoyal** [reply]
What command? If you mean jump to prompt, no all commands are sent via mappings. You can use --debug-input to see if a mapping is working.

**Reply by ldelossa** [reply]
hmm, they seem to be picked up

```
Press xkb_keycode: 0x1d clean_sym: Control_L composed_sym: Control_L mods: none glfw_key: 57442 (LEFT_CONTROL) xkb_key: 65507 (Control_L)
on_key_input: glfw key: 0xe062 native_code: 0xffe3 action: PRESS mods: none text: '' state: 0 ignoring as keyboard mode does not support encoding this event
Press xkb_keycode: 0x38 clean_sym: Alt_L composed_sym: Alt_L mods: ctrl glfw_key: 57443 (LEFT_ALT) xkb_key: 65513 (Alt_L)
on_key_input: glfw key: 0xe063 native_code: 0xffe9 action: PRESS mods: ctrl text: '' state: 0 ignoring as keyboard mode does not support encoding this event
Press xkb_keycode: 0x19 clean_sym: p composed_sym: p mods: ctrl+alt glfw_key: 112 (p) xkb_key: 112 (p)
on_key_input: glfw key: 0x70 native_code: 0x70 action: PRESS mods: ctrl+alt text: '' state: 0 
KeyPress matched action: scroll_to_prompt
handled as shortcut
Release xkb_keycode: 0x19 clean_sym: p mods: ctrl+alt glfw_key: 112 (p) xkb_key: 112 (p)
on_key_input: glfw key: 0x70 native_code: 0x70 action: RELEASE mods: ctrl+alt text: '' state: 0 ignoring release event for previous press that was handled as shortcut
Press xkb_keycode: 0x31 clean_sym: n composed_sym: n mods: ctrl+alt glfw_key: 110 (n) xkb_key: 110 (n)
on_key_input: glfw key: 0x6e native_code: 0x6e action: PRESS mods: ctrl+alt text: '' state: 0 
KeyPress matched action: scroll_to_prompt
handled as shortcut
Release xkb_keycode: 0x31 clean_sym: n mods: ctrl+alt glfw_key: 110 (n) xkb_key: 110 (n)
```

should I be seeing the "-1" and "1" arguments to "scroll_to_prompt" ?

**Reply by kovidgoyal** [reply]
You said you are using default configuration so you dont need to add anything. IIRC if you dont add them -1 is assumed.

**Comment by page-down** [comment, marked as answer]
> ... Is there any further way to debug this? ...
> ... Is there a way to see if the necessary functions exist in ZSH?

If the global rc file has special configs and there are no zsh related files under $HOME, then automatic shell integration will not be enabled.
You can explicitly load and check it manually.

Download this file:
https://github.com/kovidgoyal/kitty/blob/master/shell-integration/zsh/kitty-integration

Put `echo` or `_ksi_debug_print` here or there in the file.

Add the following to .zshrc:
```zsh
export KITTY_SHELL_INTEGRATION="enabled"
autoload -Uz -- /path/to/kitty-integration; kitty-integration; unfunction kitty-integration
```

Run and find out where it goes.
```shell
kitty --config=NONE -o shell=zsh -o shell_integration=no-rc
```

To rule out the keystroke issue, you can use remote control to scroll to prompt, but you can see in your log that it is already recognized correctly.

After confirming that this works, you can go ahead and check what went wrong with the automatic shell integration.

**Reply by ldelossa** [reply]
looks like this in my zshrc made it work:

export KITTY_SHELL_INTEGRATION="enabled"
autoload -Uz -- /usr/lib64/kitty/shell-integration/zsh/kitty-integration; kitty-integration; unfunction kitty-integration

I just found where the kitty integration file existed on my machine.

**Reply by ldelossa** [reply]
a bit of a tangent, but it seems like the feature does not work correctly when there are only two issued commands, and your left at your idle command prompt.

When there are only two commands ran, going to the previous command brings you to the first one, and going to the next command hops over the second one completely and puts you back at your idle empty prompt.

Everything works perfectly once one more command is issued (3).

**Reply by page-down** [reply]
Your symptoms make me suspect that there is a problem with some third-party linux packages.
Disclaimer: The following is speculation and I did not try it.

```
rm -r "$pkgdir"/usr/share/terminfo
rm -r "$pkgdir"/usr/lib/kitty/shell-integration
```
https://github.com/archlinux/svntogit-community/blob/8c7b294f332b8f2b87ff14888705ec764a396fd5/trunk/PKGBUILD#L45-L46

I highly doubt shell integration will work properly, along with ssh kitten.

Even with manual configuration following the official documentation, I don't think the integration scripts will load using the `KITTY_INSTALLATION_DIR` path at this point.

Not surprisingly the automatic integration does not work at all.

The following is the Linux packaging document:
https://sw.kovidgoyal.net/kitty/build/#notes-for-linux-macos-packagers

**Reply by page-down** [reply]
> ... going to the next command hops over the second one completely ...

Does the output of your second command fill the entire screen? This sounds normal, because when there is not enough content, it looks like it jumps to the idle prompt.

If you believe there is a problem, you need to bisect it and provide a minimal reproducer.

**Reply by kovidgoyal** [reply]
Ugh. Why do Linux packagers do this kind of thing. I should have remembered to ask the OP if he was using a package. Good catch @page-down 

Maybe I should add some discussion templates asking for the output from kitty's config debug. That would have shown a distro package was in use.

**Reply by ldelossa** [reply]
> > ... going to the next command hops over the second one completely ...
> 
> Does the output of your second command fill the entire screen? This sounds normal, because when there is not enough content, it looks like it jumps to the idle prompt.
> 
> If you believe there is a problem, you need to bisect it and provide a minimal reproducer.

Youre right, this is the case. I can work around this typically by just using the "page last command" shortcut.

**Reply by ldelossa** [reply]
> Your symptoms make me suspect that there is a problem with some third-party linux packages. Disclaimer: The following is speculation and I did not try it.
> 
> ```
> rm -r "$pkgdir"/usr/share/terminfo
> rm -r "$pkgdir"/usr/lib/kitty/shell-integration
> ```
> 
> https://github.com/archlinux/svntogit-community/blob/8c7b294f332b8f2b87ff14888705ec764a396fd5/trunk/PKGBUILD#L45-L46
> 
> I highly doubt shell integration will work properly, along with ssh kitten.
> 
> Even with manual configuration following the official documentation, I don't think the integration scripts will load using the `KITTY_INSTALLATION_DIR` path at this point.
> 
> Not surprisingly the automatic integration does not work at all.
> 
> The following is the Linux packaging document: https://sw.kovidgoyal.net/kitty/build/#notes-for-linux-macos-packagers

Yup, I install via DNF on Fedora, would you like me to list the rpms content? That will show where each file is being extracted to. 

Im assuming all these issues go away if I just clone kitty and install from source?

**Reply by page-down** [reply]
You can also use the official installer directly, or download from GitHub Releases, unpack, and run it.

> ... I install via DNF on Fedora ...

That Fedora package looks fine. You could try backing up that file and inserting some `echo`/`_ksi_debug_print` into the `kitty-integration` script, moving the `.zshrc` and then running `kitty --config=NONE` to troubleshoot if the automatic shell integration is working properly.

If there is still a problem, the problem probably will be in the global zsh rc file or in the Linux package. You can rule out Linux package issues by trying the official binary release. (Don't forget to put an empty .zshrc under $HOME.)

That's all the advice I can give, but if it still doesn't work and you really want to know why, you'll probably have to compile and troubleshoot it yourself.

**Reply by ldelossa** [reply]
Thanks. For now, everything is working with the aforementioned lines in my .zshrc which load it from lib64 directory. I'll leave that for now, and moving forward, i'll just install from the official installer. 

Thanks a lot for the help.

**Reply by ciscohack** [reply]
@page-down I am running kitty on MacBook M2 with zsh shell and kitty shell integration seems broken and not working. tried your suggestion but that is also not helping. Keybinding ctrl+shift-g or other not working for me...can you please help me with this
